### DISCUSSION TITLE
Dealing with ConvTranspose Layers while streamlining and converting to hw?

### DISCUSSION THREAD

**Discussion by Mounice97** [Xilinx/finn#1239] [Q&A] [answered]
Is there a specific way to do the streamlining for ConvTranspose layers on an autoencoder while working with FINN? I am currently working with an autoencoder for the purpose of reducing noise in images. At the moment, it seems that brevitas has the possibility to work with QuantConvTranspose layers and train with them. However I have faced the problem that while working with FINN, there might be some poblems in the export that might not allow me to do the complete streamlining of the file along with the conversion to hw layers. Since Im2Col only works with regular convolutional layers and later these can be easily converted into ConvolutionalInputGenerators. Has anyone had any experience on how to deal with this kind of layers?
I have heard on other forums about the possibility of turning the problem around by working with upsampling+Conv layers to substitute ConvTranspose layers, but has there been any new update that would avoid this procedure?

**Comment by fpjentzsch** [comment]
Maybe @hleblevec can help you, he contributed this PR for supporting transpsed convolutions: https://github.com/Xilinx/finn/pull/835

I believe a more efficient implementation with a modified ConvolutionInputGenerator instead of upsampling/padding is also on the roadmap.

**Comment by hleblevec** [comment, marked as answer]
Hi @Mounice97 and thanks @fpjentzsch for tagging me.
For the moment there is a support of transposed convolutions implemented as a fractionnaly-strided convolution. The streamlining and convertion to hw layers can be done through the InferPixelPaddingDeconv transformation. You can find here an example of using this transformation in a custom conversion to hw layers step: https://github.com/Xilinx/finn-examples/blob/7a672f89553882854fedb0c864272ff8f0f9975d/build/espcn/custom_steps.py#L134

It works, but it's not very hardware efficient. A more efficient implementation is currently being developped, but you can use the current implementation in the meantime. Please don't hesitate to ask for help if you encounter issues with using the transformation.

**Reply by Mounice97** [reply]
I encountered some issue with this transformation while trying to do the hardware build. At the moment I have done a personalized folding for my model and estimated the amount of memory and LUT used in it  for my target FPGA, which for the moment seems reasonable. However if i tried to do the transformation of InsertAndSetFIFODepths() or directly do the ZynqBuild it shows me the following error
```
ERROR: [BD 5-390] IP definition not found for VLNV: xilinx.com:hls:FMPadding_Pixel_hls_1:1.0 
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
Cell In[78], line 9
      6 print(target_clk_cycles_per_frame)
      8 #finn_model.transform(SetFolding())
----> 9 finn_model.transform(InsertAndSetFIFODepths(fpgapart=fpga_part))
     11 finn_folding_model_filename = '../models/finn_build/MiniCAS_FINN_folding.onnx'
     12 finn_model.save(finn_folding_model_filename)

File /opt/finn/deps/qonnx/src/qonnx/core/modelwrapper.py:140, in ModelWrapper.transform(self, transformation, make_deepcopy, cleanup)
    138 model_was_changed = True
    139 while model_was_changed:
--> 140     (transformed_model, model_was_changed) = transformation.apply(transformed_model)
    141 if cleanup:
    142     transformed_model.cleanup()

File /opt/finn/src/finn/transformation/fpgadataflow/set_fifo_depths.py:325, in InsertAndSetFIFODepths.apply(self, model)
    323 model = model.transform(PrepareIP(self.fpgapart, self.clk_ns))
    324 model = model.transform(HLSSynthIP())
--> 325 model = model.transform(CreateStitchedIP(self.fpgapart, self.clk_ns))
    326 model.set_metadata_prop("exec_mode", "rtlsim")
    328 if self.force_python_sim:
    329     # do rtlsim in Python for FIFO sizing
    330     # calculate input frequency (number of cycles for each input word)

File /opt/finn/deps/qonnx/src/qonnx/core/modelwrapper.py:140, in ModelWrapper.transform(self, transformation, make_deepcopy, cleanup)
    138 model_was_changed = True
    139 while model_was_changed:
--> 140     (transformed_model, model_was_changed) = transformation.apply(transformed_model)
    141 if cleanup:
    142     transformed_model.cleanup()

File /opt/finn/src/finn/transformation/fpgadataflow/create_stitched_ip.py:617, in CreateStitchedIP.apply(self, model)
    615         model.set_metadata_prop("wrapper_filename", wrapper_filename_alt)
    616     else:
--> 617         raise Exception(
    618             """CreateStitchedIP failed, no wrapper HDL found under %s or %s.
    619             Please check logs under the parent directory."""
    620             % (wrapper_filename, wrapper_filename_alt)
    621         )
    623 return (model, False)

Exception: CreateStitchedIP failed, no wrapper HDL found under /tmp/finn_dev_orti_ma/vivado_stitch_proj_ctwfzc0_/finn_vivado_stitch_proj.srcs/sources_1/bd/finn_design/hdl/finn_design_wrapper.v or /tmp/finn_dev_orti_ma/vivado_stitch_proj_ctwfzc0_/finn_vivado_stitch_proj.gen/sources_1/bd/finn_design/hdl/finn_design_wrapper.v.
                    Please check logs under the parent directory.
```
However, if I go to the log folder from such layer FMPadding_Pixel_hls_1, i find the following error:
```
WARNING: [Common 17-1221] Tcl app 'designutils' is out of date for this release. Please run tclapp::reset_tclstore and reinstall the app.
source run_ippack.tcl -notrace
WARNING: [Common 17-1221] Tcl app 'questa' is out of date for this release. Please run tclapp::reset_tclstore and reinstall the app.
ERROR: [Common 17-685] Unable to load Tcl app xilinx::questa
ERROR: [Common 17-69] Command failed: ERROR: [Common 17-685] Unable to load Tcl app xilinx::questa

INFO: [Common 17-206] Exiting Vivado at Tue May 20 06:53:29 2025...
ERROR: [IMPL 213-28] Failed to generate IP.
```
Sometimes in other heavier models the error is of the same kind of layer but the error states to be in another region from the same app, like:
```
ERROR: [Common 17-685] Unable to load Tcl app xilinx::xsim
ERROR: [Common 17-69] Command failed: ERROR: [Common 17-685] Unable to load Tcl app xilinx::xsim
```
Any idea of how can I fix this or what am I doing wrong about this?

**Reply by fpjentzsch** [reply]
Hi, this can be caused by a Vivado bug when running multiple instances simultaneously.

Do you have the following line in your run-docker.sh? https://github.com/Xilinx/finn/commit/a9f1898deccb74a4f8e38717c5bef00e46c9f70f

**Reply by Mounice97** [reply]
I didn't have that line and seem to have solved that error, but still I encountered the following errors:
If i use the  ZynqBuild without applying the InsertAndSetFIFODepths function, then I receive the following message:
```
ERROR: [BD 41-758] The following clock pins are not connected to a valid clock source: 
/system_ila_0/clk
/system_ila_1/clk
/system_ila_2/clk

ERROR: [Common 17-39] 'validate_bd_design' failed due to earlier errors.

    while executing
"validate_bd_design"
    (file "/tmp/finn_dev_orti_ma/vivado_zynq_proj__y3xscr6/ip_config.tcl" line 158)
INFO: [Common 17-206] Exiting Vivado at Tue May 20 09:34:54 2025...
```
Otherwise if i go first with the InsertAndSetFIFODepths function, then the new error is the following:
```
%Error: Exiting due to 3 error(s)
make: Vfinn_design_wrapper.mk: No such file or directory
make: *** No rule to make target 'Vfinn_design_wrapper.mk'.  Stop.
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Cell In[98], line 9
      6 print(target_clk_cycles_per_frame)
      8 #finn_model.transform(SetFolding())
----> 9 finn_model.transform(InsertAndSetFIFODepths(fpgapart=fpga_part))
     11 finn_folding_model_filename = '../models/finn_build/MiniCAS_FINN_folding.onnx'
     12 finn_model.save(finn_folding_model_filename)

File /opt/finn/deps/qonnx/src/qonnx/core/modelwrapper.py:140, in ModelWrapper.transform(self, transformation, make_deepcopy, cleanup)
    138 model_was_changed = True
    139 while model_was_changed:
--> 140     (transformed_model, model_was_changed) = transformation.apply(transformed_model)
    141 if cleanup:
    142     transformed_model.cleanup()

File /opt/finn/src/finn/transformation/fpgadataflow/set_fifo_depths.py:394, in InsertAndSetFIFODepths.apply(self, model)
    390     else:
    391         # convnet, two inputs are typically enough to fill entire
    392         # layer pipeline due to overlaps
    393         n_inputs = 2
--> 394     sim = verilator_fifosim(model, n_inputs)
    396 for ind, node in enumerate(fifo_nodes):
    397     maxcount_name = "maxcount_%d" % ind

File /opt/finn/src/finn/util/pyverilator.py:244, in verilator_fifosim(model, n_inputs, max_iters)
    241 launch_process_helper(make_args, proc_env=proc_env, cwd=build_dir)
    243 sim_launch_args = ["./Vfinn_design_wrapper"]
--> 244 launch_process_helper(sim_launch_args, cwd=build_dir)
    246 with open(build_dir + "/results.txt", "r") as f:
    247     results = f.read().strip().split("\n")

File /opt/finn/src/finn/util/basic.py:205, in launch_process_helper(args, proc_env, cwd)
    203 if proc_env is None:
    204     proc_env = os.environ.copy()
--> 205 with subprocess.Popen(
    206     args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=proc_env, cwd=cwd
    207 ) as proc:
    208     (cmd_out, cmd_err) = proc.communicate()
    209 if cmd_out is not None:

File /usr/lib/python3.10/subprocess.py:971, in Popen.__init__(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, user, group, extra_groups, encoding, errors, text, umask, pipesize)
    967         if self.text_mode:
    968             self.stderr = io.TextIOWrapper(self.stderr,
    969                     encoding=encoding, errors=errors)
--> 971     self._execute_child(args, executable, preexec_fn, close_fds,
    972                         pass_fds, cwd, env,
    973                         startupinfo, creationflags, shell,
    974                         p2cread, p2cwrite,
    975                         c2pread, c2pwrite,
    976                         errread, errwrite,
    977                         restore_signals,
    978                         gid, gids, uid, umask,
    979                         start_new_session)
    980 except:
    981     # Cleanup if the child failed starting.
    982     for f in filter(None, (self.stdin, self.stdout, self.stderr)):

File /usr/lib/python3.10/subprocess.py:1863, in Popen._execute_child(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, gid, gids, uid, umask, start_new_session)
   1861     if errno_num != 0:
   1862         err_msg = os.strerror(errno_num)
-> 1863     raise child_exception_type(errno_num, err_msg, err_filename)
   1864 raise child_exception_type(err_msg)

FileNotFoundError: [Errno 2] No such file or directory: './Vfinn_design_wrapper'
```
Should i change to the dev branch from FINN? Because at the moment im using the main branch for this process.

**Reply by fpjentzsch** [reply]
Yes, I would try on the dev branch first.

**Reply by Mounice97** [reply]
I have changed to the dev branch, and I'm still getting the same error for the missing connection with some clock pins. Thus, I have tried to open the created file with Vivado to make the connections manually. However, is this supposed to be the way to address this problem, or is there a way inside of FINN's functions to solve this without going directly to Vivado?

**Reply by fpjentzsch** [reply]
The error referred to the clock pins of "system_ila_*". I wonder why you have ILAs in your design. Did you enable `enable_hw_debug` in your build config? Maybe that feature is buggy and you can simply disable the ILAs.

**Reply by Mounice97** [reply]
Thanks, that seem to solve the issue, indeed, that feature seems to be buggy. Now, after adjusting some characteristics on the folding factors and the ram styles on certain nodes, I can now create a build. Just one more question, I am using the 'KV260_SOM' from the supported PYNQ boards in FINN. However, I am currently using the Kria KR260, and if I open the project in Vivado, then the project part is obviously for the Kria KV260. I can change it to the Kria KR260 and then regenerate the bitfiles with the corrected IPs. However, should I run again the MakePYNQDriver function after making such changes, or would this not work? In principle, how should I proceed to address my FPGA correctly?

**Reply by fpjentzsch** [reply]
I don't think you'll need to run MakePYNQDriver again in this situation.

You can see here what changes need to be made to add native support for a new ZYNQ device (KV260 as example): https://github.com/Xilinx/finn/pull/692/files

**Reply by Mounice97** [reply]
Thank you very much, @fpjentzsch. Now I have been able to complete the hardware build and deployment. Now It's just a matter of refining some details to reduce the number of clock cycles for my project.
